router/pkg/pubsub/kafka/kafka.go

```diff
@@ -0,0 +1,380 @@
+package kafka
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/twmb/franz-go/pkg/kerr"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/wundergraph/cosmo/router/pkg/pubsub"
+	"github.com/wundergraph/graphql-go-tools/v2/pkg/engine/datasource/pubsub_datasource"
+	"github.com/wundergraph/graphql-go-tools/v2/pkg/engine/resolve"
+	"go.uber.org/zap"
+	"sync"
+	"time"
+)
+
+var (
+	_ pubsub_datasource.KafkaConnector = (*connector)(nil)
+	_ pubsub_datasource.KafkaPubSub    = (*kafkaPubSub)(nil)
+	_ pubsub.Lifecycle                 = (*kafkaPubSub)(nil)
+
+	errClientClosed = errors.New("client closed")
+)
+
+type record struct {
+	topic string
+	data  []byte
+}
+
+type subscription struct {
+	topics  []string
+	updater resolve.SubscriptionUpdater
+	context context.Context
+}
+
+type connector struct {
+	writeClient *kgo.Client
+	opts        []kgo.Opt
+	logger      *zap.Logger
+}
+
+func NewConnector(logger *zap.Logger, opts []kgo.Opt) (pubsub_datasource.KafkaConnector, error) {
+
+	writeClient, err := kgo.NewClient(append(opts,
```

Comment:
Should we always create a write client, even though we're not publishing anything?
As an alternative, you could use `sync.Once` to lazily init it once necessary.

Reply:
Absolutely, the client should be ready as soon as possible. Instantiating it on demand could result in latency and I'd need to protect the creation through a mutex again.

Reply:
sync.Once handles all of this, but the additional latency might be an argument for the pre-creation
One thing we could consider is to not create a writing client if we don't have any publish operations configured.

Reply:
Honestly, this is negligible. A client does not have much overhead and we only have a single write client.